lib/filterx/expr-message-ref.c

```diff
@@ -31,27 +31,39 @@ typedef struct _FilterXMessageRefExpr
   NVHandle handle;
 } FilterXMessageRefExpr;
 
+static FilterXObject *
+_pull_variable_from_message(FilterXMessageRefExpr *self, FilterXEvalContext *context, LogMessage *msg)
+{
+  gssize value_len;
+  LogMessageValueType t;
+  const gchar *value = log_msg_get_value_if_set_with_type(msg, self->handle, &value_len, &t);
+  if (!value)
+    return NULL;
+
+  FilterXObject *msg_ref = filterx_message_value_new_borrowed(value, value_len, t);
+  filterx_scope_register_variable(context->scope, self->handle, FALSE, msg_ref);
+  return msg_ref;
+}
+
+/* NOTE: unset on a variable that only exists in the LogMessage, without making the message writable */
+static void
+_whiteout_variable(FilterXMessageRefExpr *self, FilterXEvalContext *context)
+{
+  filterx_scope_register_variable(context->scope, self->handle, FALSE, NULL);
```

Comment:
I don't fully understand this part, but it looks like `filterx_scope_sync_to_message()` will do the unset in a delayed manner, which is what we want.

Reply:
But only if there's an entry in the scope with a NULL value, which is what this does.